jujugui/static/gui/src/app/state/test-state.js

```diff
@@ -711,31 +711,228 @@ describe('State', () => {
       const state = new window.jujugui.State({
         baseURL: 'http://abc.com:123',
         seriesList:  ['precise', 'trusty', 'xenial'],
-        location: {href: 'foo'}
+        location: {href: 'hatch/ghost'}
       });
       sinon.stub(state,
         'generateState', () => ({ error: null, state: {new: 'state'}}));
       state.dispatch();
       assert.deepEqual(state._appStateHistory, [{new: 'state'}]);
     });
+
+    it('dispatches registered dispatchers in proper order', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123',
+        seriesList:  ['precise', 'trusty', 'xenial'],
+        location: {href: 'ghost/trusty/i/machines'}
+      });
+      let counter = 0;
+      let increment = () => counter += 1;
+      let execution = {};
+      const stub = function(state, next) {
+        execution.stub = increment();
+        next();
+      };
+      const stub2 = function(state, next) {
+        execution.stub2 = increment();
+        next();
+      };
+      const stub3 = function(state, next) {
+        execution.stub3 = increment();
+        next();
+      };
+      const stub4 = function(state, next) {
+        execution.stub4 = increment();
+        next();
+      };
+      state.register([
+        ['*', stub],
+        ['store', stub2],
+        ['*', stub3],
+        ['gui.machines', stub4]
+      ]);
+      state.dispatch();
+      assert.deepEqual(execution, {stub: 1, stub2: 4, stub3: 2, stub4: 3});
+    });
+
+    it('dispatches registered cleanup dispatchers in proper order', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123',
+        seriesList:  ['precise', 'trusty', 'xenial'],
+        location: {href: 'ghost/trusty/i/machines'}
+      });
+      let counter = 0;
+      let increment = () => counter += 1;
+      let execution = {};
+      const stub = function(state, next) {
+        execution.stub = increment();
+        next();
+      };
+      const stub2 = function(state, next) {
+        execution.stub2 = increment();
+        next();
+      };
+      const stub3 = function(state, next) {
+        execution.stub3 = increment();
+        next();
+      };
+      const stub4 = function(state, next) {
+        execution.stub4 = increment();
+        next();
+      };
+      const stub5 = function(state, next) {
+        execution.stub5 = increment();
+        next();
+      };
+      const stub6 = function(state, next) {
+        execution.stub6 = increment();
+        next();
+      };
+      state.register([
+        ['*', stub],
+        ['store', stub2, stub5],
+        ['*', stub3],
+        ['gui.machines', stub4, stub6]
+      ]);
+      state.dispatch(['store', 'gui.machines']);
+      assert.deepEqual(execution, {
+        stub5: 1, stub6: 2, stub: 3, stub3: 4, stub4: 5, stub2: 6});
+    });
+  });
+
+  describe('State.changeState()', () => {
+    it('can update state', () => {
```

Comment:
Can we test:
1. changing multiple state keys at once;
2. setting a whole new state.